### PR TITLE
[OSSFScorecard] Create scorecard badge service

### PR DESCRIPTION
**Pull request by rohankh532**
Badge service to display scores from the [OpenSSF Security Scorecard.](https://github.com/ossf/scorecard)

### CODE REVIEW

**Comment by PyvesB** [comment]
Hello @rohankh532 ! 👋🏻 

It would have been helpful to add some context about the Scorecard service, what these badges are trying to achieve, how they would be used, etc.

I've not looked at your implementation in details, but I did notice your `score (valid)` test which points to `github/org/repo`. The https://github.com/org/repo repository clearly isn't valid. The API seems to return sane-looking data regardless of what parameters are passed to it: https://api.securityscorecards.dev/projects/github/dqshsjqkhdqsjdhqsghd44/does-not-exist.json. I've also tried a bunch of repositories listed in https://github.com/ossf/scorecard/blob/main/cron/data/projects.csv, I've not yet managed to get any response that didn't have a score of 1, including for the Scorecard repository itself.

Is the API actually working? Is it documented anywhere?

**Comment by calebcartwright** [comment]
Believe #5941 was a request along the same lines.

For what it's worth, I do think many of OSSF-based metrics would make for good badges, but also agree that we need to make sure the APIs we'd be depending on are in good order as P-Y noted

**Comment by rohankh532** [comment]
> Hello @rohankh532 ! 👋🏻
> 
> It would have been helpful to add some context about the Scorecard service, what these badges are trying to achieve, how they would be used, etc.
> 
> I've not looked at your implementation in details, but I did notice your `score (valid)` test which points to `github/org/repo`. The https://github.com/org/repo repository clearly isn't valid. The API seems to return sane-looking data regardless of what parameters are passed to it: https://api.securityscorecards.dev/projects/github/dqshsjqkhdqsjdhqsghd44/does-not-exist.json. I've also tried a bunch of repositories listed in https://github.com/ossf/scorecard/blob/main/cron/data/projects.csv, I've not yet managed to get any response that didn't have a score of 1, including for the Scorecard repository itself.
> 
> Is the API actually working? Is it documented anywhere?

Scorecards is a tool that assesses a repository with several security checks and provides a score for different categories. It is trigger via a GitHub workflow, which would run every time the project was updated. We are currently incorporating signatures on scorecard results so that their authenticity can be proved. This way, if someone "claims" their project has a certain score, it can actually be verified.

We are using badges to allow users to display their scorecard score so that other users can get a sense for how vulnerable the project is. The API currently just returns 1 for every repository as we are still working on a method for computing an aggregate score. So for now it is just a placeholder, but the format will still remain the same.

**Comment by PyvesB** [comment]
Thanks for sharing the extra details! If the API is still work in progress, I suggest we put this PR on hold for the time being, and circle back to it once things are in a more final state. How does that sound?

**Comment by rohankh532** [comment]
> Thanks for sharing the extra details! If the API is still work in progress, I suggest we put this PR on hold for the time being, and circle back to it once things are in a more final state. How does that sound?

Sounds good. I'll change this PR to a draft for now.

**Comment by rohankh532** [comment]
The API is now functional, here are 3 example endpoints for which the scorecard results are given in response:
1. https://api.securityscorecards.dev/projects/github.com/rohankh532/scorecard-OIDC-test
2. https://api.securityscorecards.dev/projects/github.com/rohankh532/shields
3. https://api.securityscorecards.dev/projects/github.com/rohankh532/org-workflow-add

And the corresponding badges:
image
image
image

**Comment by calebcartwright** [comment]
Great news! Could you please rebase and make sure the CI jobs are passing?

**Comment by rohankh532** [comment]
> Great news! Could you please rebase and make sure the CI jobs are passing?

Thanks! Should be fixed now.

**Comment by rohankh532** [comment]
@calebcartwright Is there any update on this?

**Comment by PyvesB** [comment]
I'll aim at reviewing the PR this weekend. We're a small maintainer team, who dedicate their limited spare time to Shields for free. We do our best, but response times can be a little inconsistent. 😉

**Comment by rohankh532** [comment]
> I'll aim at reviewing the PR this weekend. We're a small maintainer team, who dedicate their limited spare time to Shields for free. We do our best, but response times can be a little inconsistent. 😉

No problem at all, thank you!

File: services/scorecard/scorecard.service.js
Comment (calebcartwright) [inline services/scorecard/scorecard.service.js:1]: We should use a more explicit, descriptive name that represents the actual upstream service. "Scorecard" is very generic and we can't just assume/assign that generic term to mean OSSF's score card.

Please rename the directory, files, and class accordingly

File: services/scorecard/scorecard.service.js
Comment (calebcartwright) [inline services/scorecard/scorecard.service.js:11]: Does the upstream service support branch-specific scans or is it only at the repo level?

Is there a fixed set of providers it integrates with (e.g. GitHub, GitLab, Bitbucket, etc.) or is it truly open ended?

File: services/scorecard/scorecard.service.js
Comment (calebcartwright) [inline services/scorecard/scorecard.service.js:40]: This is all related to rendering concerns and we have a very specific pattern we follow for this.

I'd encourage you to read through our [tutorial](https://github.com/badges/shields/blob/master/doc/TUTORIAL.md) on adding new badges which covers this as well as another piece you are missing which are the example listings that make the badge show up on Shields.io

File: services/scorecard/scorecard.service.js
Comment (calebcartwright) [inline services/scorecard/scorecard.service.js:33]: The scaled mapping of a numeric value to a range of colors is, as you might imagine, fairly common within Shields and we've a number of reusable helpers for such circumstances.

In this instance, I'd suggest taking a look at `colorScale` which can be imported from `color-formatters.js`, and [our code climate analysis class](https://github.com/badges/shields/blob/master/services/codeclimate/codeclimate-analysis.service.js) may serve as a helpful reference

File: services/scorecard/scorecard.service.js
Comment (calebcartwright) [inline services/scorecard/scorecard.service.js:36]: `score` should be more than sufficient here, though note that static attributes of the final badge can and should be set in the `defaultBadgeData` function (also covered in more detail in the aforementioned Tutorial)

File: services/scorecard/scorecard.tester.js
Comment (calebcartwright) [inline services/scorecard/scorecard.tester.js:10]: It's good to have a live test (see also our [guide on writing service tests](https://github.com/badges/shields/blob/master/doc/service-tests.md)), but we don't typically want to pin a live test to a specific value because that can make the test brittle. Instead, we follow a similar pattern as we do for validating response schemas and define the _type_ of result we'd expect to see, typically with another schema.

In this case I think `message: Joi.number().min(0)` would suffice instead of the explicit `4.2` value

**Review by calebcartwright — COMMENTED** [review]
Thanks for the PR! I've left feedback inline below with guidance on changes needed for next steps

File: services/scorecard/scorecard.service.js
Comment (Rohan6789) [reply services/scorecard/scorecard.service.js:11]: The scans are branch specific. Currently, it will only be integrated with GitHub, but there are future plans to expand it to GitLab and BitBucket

File: services/scorecard/scorecard.service.js
Comment (Rohan6789) [reply services/scorecard/scorecard.service.js:33]: Fixed this using color-formatters.js, thanks for the reference

File: services/scorecard/scorecard.tester.js
Comment (Rohan6789) [reply services/scorecard/scorecard.tester.js]: Fixed this, along with assuming the color value

File: services/scorecard/scorecard.service.js
Comment (rohankh532) [reply services/scorecard/scorecard.service.js]: The scans are branch specific. Currently, it will only be integrated with GitHub, but there are future plans to expand it to GitLab and BitBucket

File: services/scorecard/scorecard.service.js
Comment (rohankh532) [reply services/scorecard/scorecard.service.js]: Fixed this using color-formatters.js, thanks for the reference

File: services/scorecard/scorecard.tester.js
Comment (rohankh532) [reply services/scorecard/scorecard.tester.js]: Fixed this, along with assuming the color value

**Comment by rohankh532** [comment]
> Thanks for the PR! I've left feedback inline below with guidance on changes needed for next steps. Also, is there any publicly available documentation for the API endpoint?

Thanks for the feedback! The endpoint is publicly defined [here](https://api.securityscorecards.dev/) for now. More detailed documentation is WIP.

Also, I'm having issues with the circleci package test. I haven't changed any dependencies and it's erroring on Node 14. Can't seem to fix this.

**Comment by calebcartwright** [comment]
> Also, I'm having issues with the circleci package test. I haven't changed any dependencies and it's erroring on Node 14. Can't seem to fix this.

Don't worry about this for now as it's definitely not related to your changes (more info [here](https://github.com/badges/shields/pull/7788))

**Comment by calebcartwright** [comment]
> The endpoint is publicly defined [here](https://api.securityscorecards.dev/) for now. More detailed documentation is WIP.

Thanks but the endpoint being internet accessible is not the same thing as the endpoint being documented or made officially available. That is something that we would need before being able to move forward, or at least get some kind of official/authoritative blessing from the OSSF project. There's a number of reasons why we require a level of "official-ness" for the upstream APIs we consume, with many of them well summarized [here](https://github.com/badges/shields/issues/5506#issuecomment-1067129626) for reference

**Comment by azeemshaikh38** [comment]
> Thanks but the endpoint being internet accessible is not the same thing as the endpoint being documented or made officially available. That is something that we would need before being able to move forward, or at least get some kind of official/authoritative blessing from the OSSF project. There's a number of reasons why we require a level of "official-ness" for the upstream APIs we consume, with many of them well summarized [here](https://github.com/badges/shields/issues/5506#issuecomment-1067129626) for reference

Hi @calebcartwright, I'm one of the admins on the Scorecard project and part of the OSSF WGs. We plan to officially announce both the Scorecard API and the Scorecard badges to the public soon and I can vouch for the officialness of this project both from the Scorecard and OSSF-side. We would like to do a pre-release of these features to some OSSF repositories first and gather feedback before announcing to the larger public. So getting this PR merged would unblock us from doing that. Let me know if this helps.

**Comment by calebcartwright** [comment]
Thanks @azeemshaikh38! That should provide us sufficient cover for consuming the endpoint. 

> We would like to do a pre-release of these features to some OSSF repositories first and gather feedback before announcing to the larger public. So getting this PR merged would unblock us from doing that.

I'm not entirely sure I understood this part. Are you suggesting that our consumption of your API as part of our delivery of badges to your users is something that would be beneficial for you as part of testing/validation/feedback of your API?

**Comment by calebcartwright** [comment]
@rohankh532 you should be able to rebase now to pull in the fix for the CI issue

**Comment by rohankh532** [comment]
> @rohankh532 you should be able to rebase now to pull in the fix for the CI issue

Yup its fixed now, thanks!

**Comment by azeemshaikh38** [comment]
> I'm not entirely sure I understood this part. Are you suggesting that our consumption of your API as part of our delivery of badges to your users is something that would be beneficial for you as part of testing/validation/feedback of your API?

We are looking to test/feedback both - API and badge. They are both new features rolling out to the public. The way this works is - when you "enable" Scorecard on a repository, you get an API which provides latest Scorecard result and the ability to add a Scorecard badge to your repo. So we'll be collecting feedback on user journey, badge design etc.

**Comment by calebcartwright** [comment]
> The way this works is - when you "enable" Scorecard on a repository, you get an API which provides latest Scorecard result and the ability to add a Scorecard badge to your repo. So we'll be collecting feedback on user journey, badge design etc.

I guess that's why this is returning an http 500 response

> $ curl -i  https://api.securityscorecards.dev/projects/github.com/badges/shields

As an early piece of feedback, I'll share that's semi-problematic for us, though not necessarily blocking. We typically try to map the http errors we receive when trying to fetch data from the upstream provider to something more contextual to our badge users. However, because this is returning a 500 response we have no way of distinguishing between whether there's a legitimate upstream service issue or the user hasn't enabled their repository or if they just had a typo in the badge route parameter or something else.

**Comment by azeemshaikh38** [comment]
> As an early piece of feedback, I'll share that's semi-problematic for us, though not necessarily blocking. We typically try to map the http errors we receive when trying to fetch data from the upstream provider to something more contextual to our badge users. However, because this is returning a 500 response we have no way of distinguishing between whether there's a legitimate upstream service issue or the user hasn't enabled their repository or if they just had a typo in the badge route parameter or something else.

Ah, that's a good point. I think returning `500` is a bug really. We should return a `404` in this case instead to indicate the resource was not found. Let me know what you think. 

@rohankh532 could you update the API to ensure that we don't return a `500` error status here? If the project URL is valid and we don't have any data for it, we should return `404`.

**Comment by rohankh532** [comment]
> > As an early piece of feedback, I'll share that's semi-problematic for us, though not necessarily blocking. We typically try to map the http errors we receive when trying to fetch data from the upstream provider to something more contextual to our badge users. However, because this is returning a 500 response we have no way of distinguishing between whether there's a legitimate upstream service issue or the user hasn't enabled their repository or if they just had a typo in the badge route parameter or something else.
> 
> Ah, that's a good point. I think returning `500` is a bug really. We should return a `404` in this case instead to indicate the resource was not found. Let me know what you think.
> 
> @rohankh532 could you update the API to ensure that we don't return a `500` error status here? If the project URL is valid and we don't have any data for it, we should return `404`.

Fixed this to differentiate between `404` and `500` errors.
Ex: https://shields-staging-pr-7687.herokuapp.com/ossf-scorecard/github.com/invalid/invalid

File: services/ossf-scorecard/ossf-scorecard.service.js
Comment (calebcartwright) [inline services/ossf-scorecard/ossf-scorecard.service.js:46]: Our request handling infrastructure actually covers these cases already. We obviously allow for individual classes to apply their own message overriding for specific response codes, but it's not necessary to do so. In this case since we're not doing anything particularly unique/adding additional context I'd recommend that we remove all of these overrides, but at a minimum do so for the 500 (in order to keep it consistent and make things a tad easier/more obvious for us from an operations/support perspective)

File: services/ossf-scorecard/ossf-scorecard.service.js
Comment (calebcartwright) [inline services/ossf-scorecard/ossf-scorecard.service.js:52]: Please remove all of these

File: services/ossf-scorecard/ossf-scorecard.service.js
Comment (calebcartwright) [inline services/ossf-scorecard/ossf-scorecard.service.js:23]: Is your intent to support integrating with server/self-hosted instances of the product(s) you'll support? I ask because we've a very large number of badges that are based on the big source code platforms, but we typically don't require our users to include the TLD for those platforms.

If the OSSF service does/will support any arbitrary git host then it obviously becomes necessary to incorporate more of the uri, although if that's the case then we'd need to rethink the route a bit to account for scenarios where people have opted to host their self-hosted instance at some non-root level path, e.g. my-awesome-github-enterprise-server.com/some-path/my-org/my-repo

File: services/ossf-scorecard/ossf-scorecard.service.js
Comment (calebcartwright) [inline services/ossf-scorecard/ossf-scorecard.service.js:6]: This is a numeric, >=0 value field in the raw response correct? If so let's reflect that in that in the schema

File: services/ossf-scorecard/ossf-scorecard.tester.js
Comment (calebcartwright) [inline services/ossf-scorecard/ossf-scorecard.tester.js:17]: Please add one more case that hits the 404 path, presumably by specifying some non-existent repo

File: services/ossf-scorecard/ossf-scorecard.service.js
Comment (rohankh532) [reply services/ossf-scorecard/ossf-scorecard.service.js:46]: I removed the 500 and kept the 404

File: services/ossf-scorecard/ossf-scorecard.tester.js
Comment (rohankh532) [reply services/ossf-scorecard/ossf-scorecard.tester.js:17]: Added a test for this

File: services/ossf-scorecard/ossf-scorecard.service.js
Comment (azeemshaikh38) [reply services/ossf-scorecard/ossf-scorecard.service.js:23]: As of now, the primary audience for Scorecard badges is public repositories hosted on GitHub. Although the Scorecard team has plans of supporting non-GitHub VCS and non-public (say enterprise servers), we aren't there yet. So I don't want to overcomplicate the design here by trying to incorporate a generic uri. What if we update the uri to: `/github/:host/:repo`? If we plan to extend this in the future for other platforms, we can consider a different badge service. Would that work? @calebcartwright

Comment (calebcartwright) [reply services/ossf-scorecard/ossf-scorecard.service.js:23]: > If we plan to extend this in the future for other platforms, we can consider a different badge service. Would that work? @calebcartwright

No it would not. We don't want to be in the business of having duplicative routes/codes/etc. within Shields based on what a given provider may or may not do.

Let's just leave this as-is for now, and if/when you all make relevant changes upstream we can adjust the parameter in the route to support any optional extra paths

**Review by calebcartwright — APPROVED** [review]
Thanks for the dialog and collaboration! This looks to be in good order so going to merge :rocket:

**Comment by azeemshaikh38** [comment]
> Thanks for the dialog and collaboration! This looks to be in good order so going to merge 🚀

Thanks a lot for all your help here @calebcartwright!

**Comment by rohankh532** [comment]
> Thanks for the dialog and collaboration! This looks to be in good order so going to merge 🚀

That's great--thank you so much for your help!

**Pull request merged**
